enhancements/20240506-cpu-pinning.md

```diff
@@ -0,0 +1,219 @@
+We can examine VM CPU set by running
+```sh
+kubectl exec -n default -it virt-launcher-test-9nmwp -- taskset -cp 1
+```
+the output is
+```txt
+pid 1's current affinity list: 0-47
+```
+
+After setting up `cpu-manager-policy` to `static` on node1 and node2, we create another VM named test2 with 16 CPUs and enable CPU pinning.
+The CPU set in `test2` is `1-8,25-32` (16 CPUs). To examine the CPUs are pinned in `test2`, we can run cmd as below. 
+```sh
+kubectl exec virt-launcher-test2-7wgvk -- virsh dumpxml default_test2 | awk "/<cputune>/,/<\/cputune>/"`
+```
+the output is
+```xml
+  <cputune>
+    <vcpupin vcpu='0' cpuset='1'/>
+    <vcpupin vcpu='1' cpuset='25'/>
+    <vcpupin vcpu='2' cpuset='2'/>
+    <vcpupin vcpu='3' cpuset='26'/>
+    <vcpupin vcpu='4' cpuset='3'/>
+    <vcpupin vcpu='5' cpuset='27'/>
+    <vcpupin vcpu='6' cpuset='4'/>
+    <vcpupin vcpu='7' cpuset='28'/>
+    <vcpupin vcpu='8' cpuset='5'/>
+    <vcpupin vcpu='9' cpuset='29'/>
+    <vcpupin vcpu='10' cpuset='6'/>
+    <vcpupin vcpu='11' cpuset='30'/>
+    <vcpupin vcpu='12' cpuset='7'/>
+    <vcpupin vcpu='13' cpuset='31'/>
+    <vcpupin vcpu='14' cpuset='8'/>
+    <vcpupin vcpu='15' cpuset='32'/>
+  </cputune>
+```
+
+Subsequently, when we check VM `test`, we observe that the CPU set in it changes to `0,9-24,33-47`. The CPUs occupied by `test2`
+are excluded from the CPU shared pool.
+
+Now, let's examine other pre-existing pod, such as the `harvester-node-manager` pod deployed on node1. The CPU set in it is also `0,9-24,33-47`.
+
+#### Story 2: Set cpu-manager-policy from static to none
+Assume that we have 3 nodes (node1, node2, node3), each node contains cpu set `0-47`, and the `cpu-manager-policy` is set to `static` in node1, node2.
+Initially, we have a VM named `test` with CPU pinning enabled, utilizing 16 CPUs(`1-8,25-32`), deployed to node1.
+
+Now, let's change the `cpu-manager-policy` to `none` in node1, and observe the CPU set in VM `test`, which remains `1-8,25-32`.
+Next, we create another VM named `test2` with 2 CPUs without enabling CPU pinning and deploy it to node1,
+the CPU set in VM `test2` is `0-47`. Although CPU pinning in `test` still work, CPUs inside `test` is no longer isolated
+and could be shared with other VMs. 
+
+Upon checking other pre-existing pod, such as the `harvester-node-manager` pod which deployed in node1, we find that
+the cpu set in it remains `0,9-24,33-47`, despite the `cpu-manager-policy` being set to none in node1.
+
+Now, if we restart VM `test`, it will be deployed to node2 since `cpu-manager-policy` is set to `none` in node1, node3.
+Subsequently, if we disable CPU pinning in node2 and restart VM test, it won't deploy successfully since no node has set
+`cpu-manager-policy` to `static`.
+
+#### Story 3: users who don't want to deploy workload to nodes that set cpu-manager-policy to static
```

Comment:
When a node with `cpu-manager-policy` static and has already deployed many VMs, e.g. 80% CPUs are allcated/used, then user add a pin'ed VM to this node, what will happen? will the new VM fail to deploy (say only 1 node pin-able), or some VMs are migrated to other nodes to give space for it?

Reply:
In such scenario, the new VM will fail to deploy. And raise error like `0/1 nodes are available: 1 Insufficient cpu...`.